src/Semantics.js

```diff
@@ -196,6 +196,70 @@ Semantics.prototype.checkActionDicts = function() {
   }
 };
 
+Semantics.prototype.toRecipe = function(semanticsOnly) {
+  function hasSuperSemantics(s) {
+    return s.super !== Semantics.BuiltInSemantics._getSemantics();
+  }
+
+  var str = '(function(g) {\n';
+  if (hasSuperSemantics(this)) {
+    str += '  var semantics = ' + this.super.toRecipe(true) + '(g';
+
+    var superSemanticsGrammar = this.super.grammar;
+    var relatedGrammar = this.grammar;
+    while (relatedGrammar !== superSemanticsGrammar) {
+      str += '.superGrammar';
+      relatedGrammar = relatedGrammar.superGrammar;
+    }
+
+    str += ');\n';
+    str += '  return g.extendSemantics(semantics)';
+  } else {
+    str += '  return g.semantics()';
+  }
+  ['Operation', 'Attribute'].forEach(function(type) {
+    var semanticOperations = this[type.toLowerCase() + 's'];
+    Object.keys(semanticOperations).forEach(function(name) {
+      var signature = name;
+      if (semanticOperations[name].formals.length > 0) {
+        signature += '(' + semanticOperations[name].formals.join(', ') + ')';
+      }
+
+      var method;
+      if (hasSuperSemantics(this) && this.super[type.toLowerCase() + 's'][name]) {
+        method = 'extend' + type;
+      } else {
+        method = 'add' + type;
+      }
+      str += '\n    .' + method + '(' + JSON.stringify(signature) + ', {';
+
+      var actions = semanticOperations[name].actionDict;
+      var defaultStr = newDefaultAction(/* can omit params */).toString();
+      var srcArray = Object.keys(actions).filter(function(actionName) {
```

Comment:
IMO, this would be slightly clearer as a `forEach` with an `if` statement inside, but that's just a personal preference.